PopupMenu rework and enhancements.

EricEzaM opened the pull request:
Initially inspired by [this](https://www.reddit.com/r/godot/comments/id93w6/resizing_the_optionbutton_popupmenu/) reddit post, which never got it's own issue submitted. This PR resolves a few existing issues with PopupMenu's which did have issue tickets, but also a few others which I found when making this PR.

* Popup menu size can now be limited, and content will scroll. Previously, if the popup menu had a lot of items, it would take up the entire window height and then scroll from there. Now, the size of popup menus can be limited by using `get_popup()->set_max_size()` (in C++) or `get_popup().max_size` (in GDScript). Of course, this works both in editor and in projects.
* This can be selectively used around the editor to help resolve ui issues such as #22404. Examples in-editor shown below.
![pum_in_game](https://user-images.githubusercontent.com/41730826/91675099-7e901580-eb7e-11ea-8167-9a6c4f2a5035.gif)
* Added margin container which improves consistency of look, as well as makes item positioning code simpler and easier to understand.
* Now uses a scroll container so users know it can be scrolled. This also allows the user to scroll without a scroll wheel by clicking and dragging the scrollbar. Previously users needed a scroll wheel to be able to scroll. (Closes #29111)
![pum_scrolling](https://user-images.githubusercontent.com/41730826/91675117-9071b880-eb7e-11ea-8126-5b806b5cbd61.gif)
* Changing item selection with arrow keys now scrolls to follow selected item, whereas previously it did not. (resolves this comment by @KoBeWi, which did not have a separate issue: https://github.com/godotengine/godot/issues/29111#issuecomment-495631414)
* Scrolling with arrow keys stops at first or last item when scrolling up and down, respectively. Previously it looped around and kept on going. 
![pum_arrow_scrolling](https://user-images.githubusercontent.com/41730826/91675132-9c5d7a80-eb7e-11ea-88cf-f0529043b4cf.gif)
* Fixed submenu autohide areas to match exactly the hover area. Previously they were slightly off. Also fixed selection of which item was hovered, as that was also slightly off. Previously the darker area which represented which item was hovered did not accurately represent the size of the actual item. As you can see, submenu's are correctly placed when scrolling is used. Note: I am not recommending that the project menu is made to scroll, this was only for submenu testing purposes.
![pum_submenus](https://user-images.githubusercontent.com/41730826/91675151-a7180f80-eb7e-11ea-9c32-2a2683e80b61.gif)
* When search is enabled, selected item from search query will be scrolled to and highlighted. In the below case I type "obj" and then "res".
![pum_scroll_to_search](https://user-images.githubusercontent.com/41730826/91675189-cadb5580-eb7e-11ea-8a38-ed20091da011.gif)
* Reduced code duplication by moving query of item icon size to a member function in the `Item` struct.
* General code clean up - comments where needed, and rewording variables to be more clear about what they are.

*Bugsquad edit:* Also fixes #41564.

EricEzaM committed:
PopupMenu rework and enhancements

Many scrolling behaviour improvements and the ability to limit popup size.

KoBeWi commented:
Would be cool if you added [hysteresis](https://github.com/godotengine/godot/issues/11219) too 🙃
I have an unfinished PR that attempted to implement it: #35487

EricEzaM commented:
@KoBeWi That's a good feature, I'll see if I can add it in this PR.

groud commented:
It looks great ! :) I'll have a look to the code.

EricEzaM commented:
@KoBeWi I have made some progress on adding that, but I think I would put it in a separate PR so it can be reviewed separately from this PR, and not too many features/changes get muddled up in one PR.

groud reviewed (approved):
The code looks good to me, but i haven't tested locally.

KoBeWi reviewed (changes requested):
I did some testing and I noticed that some popups have maximum size that don't really make sense
![VXpMW3yhhJ](https://user-images.githubusercontent.com/2223172/91724546-beb2c080-eb9d-11ea-9a15-72e6d62938fc.gif)
Looks like it's calculated wrongly or something.

~~Also the new `max_size` property needs documentation.~~ (I now noticed it comes from Window)

Other than that, all seems to work fine.

EricEzaM commented:
Hmm yeah it looks like the height of the contents  is calculated with an extra very small value, so the contents has to scroll a tiny bit. Will investigate in the morning

EricEzaM commented:
Looks like there are a number of issues to resolve...

- [x] 1. (This existed before this PR) Popups have height set to 100px even if they do not need it. See `File -> Convert To...`. The weird thing is that this doesn't seem to apply to all menus... See `File -> Open Recent` (when the list is empty), which does not have the same issue. Edit: Resolved

![image](https://user-images.githubusercontent.com/41730826/91779465-f91d6b80-ec38-11ea-9d9f-eaa0b955fe9a.png)
![image](https://user-images.githubusercontent.com/41730826/91779471-fe7ab600-ec38-11ea-82ce-06123d3fd5e4.png)

- [x] 2. The issue Kobewi pointed out above, plus: the scroll bars do not focus automatically on what seems to be only these popup menus. When you hover the bar it does not highlight. Additionally, when you do click on the scrollbar *then* is highlights and you can click and scroll. Also, if you close it in it's highlighted state, it will stay that way when it comes up again. So it seems like the scrollbar is frozen to input or something? very weird. Edit: Resolved

![popup_bug](https://user-images.githubusercontent.com/41730826/91780210-38988780-ec3a-11ea-93a0-5aa79fc9545e.gif)


- [x] 3. When a max size is not set, and the popup menu minimum size returns something higher than the screen height, no scrollbar is visible and you cannot scroll. Edit: Resolved now.

![image](https://user-images.githubusercontent.com/41730826/91779667-92e51880-ec39-11ea-94f0-d6d3725f11bb.png)

EricEzaM commented:
@KoBeWi ok, all done. I found some more stuff as in my previous comment which is now resolves.

The issue you found was really weird... It seems like the setting of the margin was not occurring properly on the margin container. So, the margins were 10px, but the content thought it was only 8px (which is what it should be). Therefore there was a 4px scroll (10px*2 - 8px*2). Anyway, moving the setting of the margins away from notification enter tree is what worked.

@groud if you care to re-review, the changes of the most recent commit (which were squashed into one) were:

`popup.cpp`

- Added check which ensures popup is not larger than parent, and also does an early exit is max size is not set (Lines 129 to 141). This is part of what solved `#3` above

`popup_menu.cpp`

- Limited PopupMenu height to usable parent rect height. (lines 104 to 108) - this fixed `#3` above
- Used `set_as_minsize()` to 'shrink' the borders of the submenu (line 200) - this is what fixed `#1` above. Also did this in the `PopupMenu::Popup()` method, in case it is not a submenu.
- Moved setting of margin overrides for the margin container from ENTER_TREE to VISIBILITY_CHANGED (and after `is_visible()` check) - fixed Kobewi's issue and `#2` above

EricEzaM commented:
As a bonus, this PR also fixes the issue where popup menus were going off screen:
![image](https://user-images.githubusercontent.com/41730826/91789935-ffb8dc80-ec52-11ea-933d-067cc0c0ebac.png)

After:
![popup_in_screen](https://user-images.githubusercontent.com/41730826/91790353-e4020600-ec53-11ea-9f17-96fcd48c06c8.gif)

KoBeWi reviewed (approved):
Looks fine now after brief testing.

pouleyKetchoupp commented on scene/gui/popup.cpp:
```diff
@@ -126,6 +126,27 @@ Rect2i Popup::_popup_adjust_rect() const {
+	}
+
+	// Early out if max size not set.
+	if (get_max_size() <= Size2()) {
```
You could call get_max_size() only once and keep the value locally here. Since it's doing a platform specific call, that would ensure it never does anything expensive several times.

EricEzaM replied:
Yeah good point, calling so many `get` methods in the same scope is quite the code smell.

pouleyKetchoupp commented on scene/gui/popup_menu.h:
```diff
@@ -57,11 +59,17 @@ class PopupMenu : public Popup {
 		bool shortcut_is_disabled;
 
+		// Returns (0,0) if icon is null.
+		const Size2 get_icon_size() const {
```
The first const is not needed for a return value by copy.

pouleyKetchoupp reviewed (commented):
Great work! I've just commented on some details.

I can also confirm the latest changes fix #41564.

akien-mga commented:
Thanks! Amazing work :)

groud commented:
Awesome, thanks a lot !

chiguireitor commented:
will this be backported to the v3 branch?

SJcoded commented:
Is this obselete?

Calinou commented:
> Is this obselete?

This pull request was merged in the `master` branch (for 4.0.alpha), but it won't be cherry-picked to `3.x` as it introduces compatibility-breaking changes.

Ev1lbl0w commented:
I'm interested in backporting these changes, but what are the mentioned "breaking changes"? And what features should I avoid backporting?

EricEzaM commented:
I'm not too sure. @Calinou is it breaking change because of the fact that it changes the way the popup looks?

I don't think there was any breaking api changes by looking at the code, only the addition of the Max_size property. If you like, you can try backporting it and see how you go. Working through it will probably make any breaking changes more obvious.
